features/entire-workflow-versioning.md

```diff
@@ -0,0 +1,68 @@
+# Entire Workflow Versioning
+
+### Problem statement
+
+The current method of versioning Workflows is complex and error prone.
+By default, there's a good chance that upgrading a worker will result in workflows being stuck due to non-determinism errors.
+Recovering from this situation is not trivial because a user now has histories generated by mulitple versions.
+
+### Proposal
+
+Entire Workflow versioning is an approach that'll greatly simplify the current Workflow versioning situation.
+
+- We add a concept of a central repository for Workflow code
+- Users deploy their Workflows directly to the central repository
+- When a worker encounters the first Workflow Task in history it decides which Workflow code it needs to run
```

Comment:
Why not have the server decide which version to run, and include it in the workflowexecutionstarted event? Then there could be additional options to the startworkflowexecution rpc to choose a specific version (or "testing" tag, etc.)  Also it could help ensure that a batch of workflows are all started with the same version, even during a deployent.

Reply:
This touches worker deployment which we decided to avoid for the sake of narrowing the scope of this proposal.
Setting the version from the server is an option, it adds more work for the server while this feature might not be supported by all SDKs.
Having the tag set in the startworkflowexecution request is something we'd have to consider.

I think for now we should assume there's only a `latest` tag.